README.md
# bw_migrations

Migration data and utilities for Brightway IO and LCA in general

[![PyPi Version](https://img.shields.io/pypi/v/bw_migrations.svg)](https://pypi.org/project/bw_migrations/) [![Conda Version](https://img.shields.io/conda/vn/conda-forge/bw_migrations.svg)](https://anaconda.org/conda-forge/bw_migrations) [![Build Status](https://travis-ci.org/brightway-lca/bw_migrations.svg?branch=master)](https://travis-ci.org/brightway-lca/bw_migrations) [![Coverage Status](https://coveralls.io/repos/github/brightway-lca/bw_migrations/badge.svg?branch=master)](https://coveralls.io/github/brightway-lca/bw_migrations?branch=master) [![Build status](https://ci.appveyor.com/api/projects/status/lk0tbo21v2irm48x?svg=true)](https://ci.appveyor.com/project/cmutel/bw-migrations)

Most databases use their own nomenclature for classification systems, units, etc. These systems need to be matched when linking from one database to another. Often, a simple mapping is suitable, and tools like [correspondentia](https://github.com/BONSAMURAIS/correspondentia) are a good fit. However, sometimes one needs more complexity, e.g. change field X to Y, but only if field A has value B. When ecoinvent released version 3, they changed their unit of mesaure for water from kilograms to cubic meters. ``bw_migrations`` provides tools for this more complicated transformations, and is built around the following data format:

    {
        # The fields on which to filter
        'fields': ['name', 'category', 'unit'],
        'data': [
            (
                # First element is input data in the order of `fields` above
                ('Water', 'air', 'kilogram'),
                # Second element is new values to substitute when all fields match
                {
                    'unit': 'cubic meter',
                    '__multiplier__': 0.001
                }
            )
        ]
    }

And is implemented with the following pseudo-code:

    for element in input_data:
        for original, new in migration['data']:
            if all(element[field] == original[field] for field in migration['fields']):
                element.update(dict(zip(migration['fields'], new)))

The actual code is a bit more complex, as `bw_migrations` can also do rescaling of probability distributions and disaggregation migrations (splitting one object into several outputs).


CHANGES.md
# migrations Changelog

## 0.2 (2021-10-01)

Added migrations for EXIOBASE to ecoinvent

## 0.1 (2019-10-29)

Current working version


tests/strategies.py
from bw_migrations.strategies import modify_object, migrate_data, get_migration, DATA_DIR


def test_get_migration():
    assert get_migration({'foo': 'bar'}) == {'foo': 'bar'}
    assert len(get_migration("exiobase-3-ecoinvent-3.6")['contributors']) == 2
    assert len(get_migration(DATA_DIR / "exiobase-3-ecoinvent-3.6.json")['contributors']) == 2
    assert len(get_migration(str(DATA_DIR / "exiobase-3-ecoinvent-3.6.json"))['contributors']) == 2


def test_modify_object():
    given = {"a": 5, "b": "foo"}
    dct = {"b": "zany", "c": "wild"}
    expected = {"a": 5, "b": "zany", "c": "wild"}
    assert modify_object(given, dct) == expected
    assert given == expected


def test_modify_object_with_disaggregation():
    given = {"a": 5, "b": "foo", "amount": 11}
    dct = {"b": "zany", "c": "wild", "__disaggregation__": 5}
    expected = {"a": 5, "b": "zany", "c": "wild", "amount": 55}
    assert modify_object(given, dct) == expected
    assert given == expected


def test_modify_object_with_multiplier():
    given = {"a": 5, "b": "foo", "amount": 11}
    dct = {"b": "zany", "c": "wild", "__multiplier__": 5}
    expected = {"a": 5, "b": "zany", "c": "wild", "amount": 55}
    assert modify_object(given, dct) == expected
    assert given == expected


def test_modify_object_with_both():
    given = {"a": 5, "b": "foo", "amount": 11}
    dct = {"b": "zany", "c": "wild", "__multiplier__": 5, "__disaggregation__": 0.1}
    expected = {"a": 5, "b": "zany", "c": "wild", "amount": 5.5}
    assert modify_object(given, dct) == expected
    assert given == expected


def test_migrate_data():
    migration = {
        "fields": ["a", "b"],
        "data": [(("green", "brown"), {"a": "blue", "another": "value"})],
    }
    given = [
        {"a": 5, "b": "foo"},
        {"a": "green", "b": "foo"},
        {"a": "green", "b": "brown"},
    ]
    expected = [
        {"a": 5, "b": "foo"},
        {"a": "green", "b": "foo"},
        {"a": "blue", "b": "brown", "another": "value"},
    ]
    assert list(migrate_data(given, migration)) == expected


def test_migrate_data_disaggregation():
    migration = {
        "fields": ["a", "b"],
        "data": [
            (
                ("green", "brown"),
                [
                    {
                        "a": "blue",
                        "another": "value",
                        "__multiplier__": 0.5,
                        "__disaggregation__": 0.3,
                    },
                    {"a": "red", "__disaggregation__": 0.7},
                ],
            )
        ],
    }
    given = [
        {"a": 5, "b": "foo"},
        {"a": "green", "b": "foo"},
        {"a": "green", "b": "brown", "amount": 10},
    ]
    expected = [
        {"a": 5, "b": "foo"},
        {"a": "green", "b": "foo"},
        {"a": "blue", "b": "brown", "another": "value", "amount": 1.5},
        {"a": "red", "b": "brown", "amount": 7},
    ]
    assert list(migrate_data(given, migration)) == expected


def test_migrate_data_multiplier():
    migration = {
        "fields": ["a", "b"],
        "data": [
            (
                ("green", "brown"),
                {"a": "blue", "another": "value", "__multiplier__": 0.5},
            )
        ],
    }
    given = [
        {"a": 5, "b": "foo"},
        {"a": "green", "b": "foo"},
        {"a": "green", "b": "brown", "amount": 4},
    ]
    expected = [
        {"a": 5, "b": "foo"},
        {"a": "green", "b": "foo"},
        {"a": "blue", "b": "brown", "another": "value", "amount": 2},
    ]
    assert list(migrate_data(given, migration)) == expected


tests/test_utils.py
from bw_migrations.utils import rescale_object, get_uncertainty_type
import pytest


def test_get_uncertainty_type():
    assert get_uncertainty_type({"uncertainty type": 1}) == 1
    assert get_uncertainty_type({"uncertainty_type": 2}) == 2
    assert get_uncertainty_type({"uncertainty_type_id": 3}) == 3
    assert get_uncertainty_type({"foo": "bar"}) == 0


def test_rescale_object_number_error():
    with pytest.raises(ValueError):
        rescale_object({}, "nope")


def test_rescale_object_dist_error():
    with pytest.raises(ValueError):
        rescale_object({"uncertainty type": 42}, 7)


def test_rescale_object_formula():
    given = {"uncertainty type": 0, "formula": "hi mom", "amount": 7}
    expected = {"uncertainty type": 0, "formula": "(hi mom) * 7", "amount": 49}
    assert rescale_object(given, 7) == expected


def test_no_uncertainty():
    given = {"uncertainty type": 0, "amount": 42}
    expected = {"uncertainty type": 0, "amount": 21}
    assert rescale_object(given, 0.5) == expected


def test_no_uncertainty_with_loc():
    given = {"uncertainty type": 0, "amount": 42, "loc": "dummy"}
    expected = {"uncertainty type": 0, "amount": 21, "loc": 21}
    assert rescale_object(given, 0.5) == expected


def test_unknown_uncertainty():
    given = {"uncertainty type": 1, "amount": 42}
    expected = {"uncertainty type": 1, "amount": 21}
    assert rescale_object(given, 0.5) == expected


def test_unknown_uncertainty_with_loc():
    given = {"uncertainty type": 1, "amount": 42, "loc": "dummy"}
    expected = {"uncertainty type": 1, "amount": 21, "loc": 21}
    assert rescale_object(given, 0.5) == expected


def test_normal_uncertainty():
    given = {"uncertainty type": 3, "amount": 2, "scale": 4}
    expected = {"uncertainty type": 3, "amount": 3, "loc": 3, "scale": 6}
    assert rescale_object(given, 1.5) == expected


def test_lognormal_uncertainty():
    given = {"uncertainty type": 2, "amount": 2, "loc": 2, "scale": 4}
    expected = {"uncertainty type": 2, "amount": 3, "loc": 3, "scale": 4}
    assert rescale_object(given, 1.5) == expected


def test_uniform_uncertainty():
    given = {"uncertainty type": 4, "amount": 2, "minimum": 0, "maximum": 4}
    expected = {"uncertainty type": 4, "amount": 3, "minimum": 0, "maximum": 6}
    assert rescale_object(given, 1.5) == expected


def test_uniform_uncertainty_flip_sign():
    given = {"uncertainty type": 4, "amount": 2, "minimum": 0, "maximum": 4}
    expected = {"uncertainty type": 4, "amount": -3, "minimum": -6, "maximum": 0}
    assert rescale_object(given, -1.5) == expected


def test_uniform_uncertainty_flip_sign_already_negative():
    given = {"uncertainty type": 4, "amount": -2, "minimum": -4, "maximum": 0}
    expected = {"uncertainty type": 4, "amount": 3, "minimum": 0, "maximum": 6}
    assert rescale_object(given, -1.5) == expected


def test_uniform_uncertainty_no_amount_field():
    given = {"uncertainty type": 4, "minimum": 0, "maximum": 4}
    expected = {"uncertainty type": 4, "minimum": 0, "maximum": 6}
    assert rescale_object(given, 1.5) == expected


def test_triangular_uncertainty():
    given = {"uncertainty type": 5, "amount": 2, "minimum": 0, "maximum": 4}
    expected = {
        "uncertainty type": 5,
        "amount": 3,
        "minimum": 0,
        "maximum": 6,
        "loc": 3,
    }
    assert rescale_object(given, 1.5) == expected


def test_triangular_uncertainty_flip_sign():
    given = {"uncertainty type": 5, "amount": 2, "minimum": 0, "maximum": 4}
    expected = {
        "uncertainty type": 5,
        "amount": -3,
        "minimum": -6,
        "maximum": 0,
        "loc": -3,
    }
    assert rescale_object(given, -1.5) == expected


def test_triangular_uncertainty_flip_sign_already_negative():
    given = {"uncertainty type": 5, "amount": -2, "minimum": -4, "maximum": 0}
    expected = {
        "uncertainty type": 5,
        "amount": 3,
        "minimum": 0,
        "maximum": 6,
        "loc": 3,
    }
    assert rescale_object(given, -1.5) == expected


bw_migrations/__init__.py
from .version import version as __version__

__all__ = (
    "get_migration",
    "migrate_data",
)

from .strategies import migrate_data, get_migration


# def create_core_migrations():
#     """Add pre-defined core migrations data files"""
#     Migration("biosphere-2-3-categories").write(
#         get_biosphere_2_3_category_migration_data(),
#         "Change biosphere category and subcategory labels to ecoinvent version 3"
#     )
#     Migration("biosphere-2-3-names").write(
#         get_biosphere_2_3_name_migration_data(),
#         "Change biosphere flow names to ecoinvent version 3"
#     )
#     Migration("simapro-ecoinvent-3.1").write(
#         get_simapro_ecoinvent_3_migration_data("3.1"),
#         "Change SimaPro names from ecoinvent 3.1 to ecoinvent names"
#     )
#     Migration("simapro-ecoinvent-3.2").write(
#         get_simapro_ecoinvent_3_migration_data("3.2"),
#         "Change SimaPro names from ecoinvent 3.2 to ecoinvent names"
#     )
#     Migration("simapro-ecoinvent-3.3").write(
#         get_simapro_ecoinvent_3_migration_data("3.3"),
#         "Change SimaPro names from ecoinvent 3.3 to ecoinvent names"
#     )
#     Migration("simapro-water").write(
#         get_simapro_water_migration_data(),
#         "Change SimaPro water flows to more standard names"
#     )
#     Migration("us-lci").write(
#         get_us_lci_migration_data(),
#         "Fix names in US LCI database"
#     )
#     Migration("default-units").write(
#         get_default_units_migration_data(),
#         "Convert to default units"
#     )
#     Migration("unusual-units").write(
#         get_unusual_units_migration_data(),
#         "Convert non-Ecoinvent units"
#     )
#     Migration("exiobase-biosphere").write(
#         get_exiobase_biosphere_migration_data(),
#         "Change biosphere flow names to ecoinvent version 3"
#     )
#     Migration("fix-ecoinvent-flows-pre-35").write(
#         get_ecoinvent_pre35_migration_data(),
#         "Update new biosphere UUIDs in Consequential 3.4"
#     )


bw_migrations/strategies.py
from .utils import rescale_object
from collections.abc import Mapping
from copy import deepcopy
from pathlib import Path
import json

DATA_DIR = Path(__file__, "..").resolve() / "data"


def get_migration(location):
    if isinstance(location, Mapping):
        return location
    elif isinstance(location, Path) and Path(location).is_file():
        return json.load(open(location))
    elif isinstance(location, str) and Path(location).is_file():
        return json.load(open(location))
    elif isinstance(location, str) and any((location + ".json") in o.name for o in DATA_DIR.iterdir()):
        return json.load(open(DATA_DIR / (location + ".json")))
    else:
        raise ValueError


def modify_object(obj, dct):
    scale = dct.get("__disaggregation__", 1) * dct.get("__multiplier__", 1)
    if scale != 1:
        obj = rescale_object(obj, scale)
    for k, v in dct.items():
        if k not in ("__disaggregation__", "__multiplier__"):
            obj[k] = v
    return obj


def migrate_data(data, migration):
    migration_data = get_migration(migration)
    lookup = {tuple(x): y for x, y in migration_data["data"]}

    for row in data:
        try:
            new = lookup[tuple([row.get(field) for field in migration_data["fields"]])]
            if isinstance(new, list):
                for dct in new:
                    yield modify_object(deepcopy(row), dct)
            else:
                yield modify_object(row, new)
        except KeyError:
            yield row


bw_migrations/data/units.py
DEFAULT_UNITS_CONVERSION = [
    # Energy
    ("PJ", "megajoule", 1e9),
    ("J", "megajoule", 1e-06),
    ("Nm3 Europe", "megajoule", 37.5),
    ("Bt", "megajoule", 0.001055696),
    ("kilojoule", "megajoule", 1e-3),
    ("kg-NL", "megajoule", 42.4),
    ("Nm3-Russia", "megajoule", 36.4),
    ("Nm3-DZ", "megajoule", 38.5),
    ("Nm3-GB", "megajoule", 37.0),
    ("Nm3 - NG", "megajoule", 38.5),
    ("Nm3- Italy", "megajoule", 33.85),
    ("Nm3 F-v6", "megajoule", 36.894),
    ("kg-DZ", "megajoule", 49.7),
    ("kWp", "megajoule", 3.6),
    ("kg-DE", "megajoule", 45.9),
    ("gigajoule", "megajoule", 1e3),
    ("kcal", "megajoule", 0.0041855),
    ("kg-RU", "megajoule", 49.3),
    ("kg-NO", "megajoule", 56.6),
    ("TJ", "megajoule", 1e6),
    ("Nm3-DE", "megajoule", 35.0),
    ("Nm3 FR", "megajoule", 37.09),
    ("Nm3 - EG", "megajoule", 38.5),
    ("Nm3-NL", "megajoule", 34.9),
    ("kg-GB", "megajoule", 47.0),
    ("NM3-Norway", "megajoule", 40.8),
    ("cal", "megajoule", 4.19e-06),
    ("MWh", "kilowatt hour", 1e3),
    ("TWh", "kilowatt hour", 1e9),
    ("watt hour", "kilowatt hour", 1e-3),
    ("GWh", "kilowatt hour", 1e6),
    # Mass
    ("gram", "kilogram", 1e-3),
    ("t", "kilogram", 1e3),
    ("tMS", "kilogram", 1e3),
    ("lb", "kilogram", 0.4535924),
    ("T", "kilogram", 1e3),
    ("tn.sh", "kilogram", 907.1848),
    ("tMB", "kilogram", 1e3),
    ("pg", "kilogram", 1e-15),
    ("mg", "kilogram", 1e-06),
    ("kton", "kilogram", 1e6),
    ("oz", "kilogram", 0.02834952),
    ("?g", "kilogram", 1e-09),  # SimaPro weirdness
    ("μg", "kilogram", 1e-09),
    ("tn.lg", "kilogram", 1016.047),
    ("tOM", "kilogram", 1e3),
    ("Mtn", "kilogram", 1e9),
    ("ton", "kilogram", 1e3),
    ("ng", "kilogram", 1e-12),
    # Length
    ("yard", "meter", 0.9144),
    ("inch", "meter", 0.0254),
    ("mm", "meter", 1e-3),
    # ('kilometer', 'meter', 1e3),
    ("dm", "meter", 0.1),
    ("Nautic mil", "meter", 1547.896),
    ("mile", "meter", 1609.35),
    ("cm", "meter", 0.01),
    ("ft", "meter", 0.3048),
    ("?m", "meter", 1e-06),
    ("μm", "meter", 1e-06),
    # Length * time
    ("miy", "meter-year", 1609.35),
    ("kmy", "meter-year", 1e3),
    # Volume
    ("dm3", "cubic meter", 1e-3),
    ("cu.in", "cubic meter", 1.638706e-05),
    ("cuft", "cubic meter", 0.02831685),
    ("mm3", "cubic meter", 1e-09),
    ("gal*", "cubic meter", 0.003785412),  # SimaPro weirdness
    ("cu.yd", "cubic meter", 0.7645549),
    ("cm3", "cubic meter", 1e-6),
    # ('litre', 'cubic meter', 1e-3),
    # Volume * time
    ("l*day", "cubic meter-year", 2.7397e-06),
    ("cm3y", "cubic meter-year", 1e-06),
    ("m3day", "cubic meter-year", 0.0027397),
    # Area
    ("sq.mi", "square meter", 2589988.0),
    ("sq.ft", "square meter", 0.09290304),
    ("sq.in", "square meter", 0.00064516),
    ("sq.yd", "square meter", 0.8361273),
    ("cm2", "square meter", 0.0001),
    ("km2", "square meter", 1e6),
    ("dm2", "square meter", 0.01),
    ("acre", "square meter", 4046.856),
    ("mm2", "square meter", 1e-06),
    # Area * time
    ("mm2a", "square meter-year", 1e-06),
    ("ha a", "square meter-year", 1e4),
    ("cm2a", "square meter-year", 1e-4),
    ("km2a", "square meter-year", 1e6),
    # Radiation
    ("nBq", "kilo Becquerel", 1e-12),
    ("Becquerel", "kilo Becquerel", 1e-3),
    ("MBq", "kilo Becquerel", 1e3),
    ("?Bq", "kilo Becquerel", 1e-9),  # SimaPro weirdness
    ("μBq", "kilo Becquerel", 1e-9),
    ("mBq", "kilo Becquerel", 1e-6),
    # Time
    ("s", "hour", 3600.0),
    ("day", "hour", 1 / 24),
    ("year", "hour", 1 / 8765.81),
    ("min", "hour", 60.0),
    # Sector-specific
    ("pmi", "person kilometer", 1.60935),
    ("kilogram kilometer", "ton kilometer", 1e-3),
    ("tmi*", "ton kilometer", 1.45997),
    ("ktkm", "ton kilometer", 1e3),
]


def get_default_units_migration_data():
    return {
        "fields": ["unit"],
        "data": [
            ((from_unit,), {"unit": to_unit, "multiplier": multiplier})
            for from_unit, to_unit, multiplier in DEFAULT_UNITS_CONVERSION
        ],
    }


_USED_IN_ECOINVENT = {
    "cubic meter",
    "hectare",
    "hour",
    "kg SWU",
    "kilogram",
    "kilometer",
    "kilometer-year",
    "kilowatt hour",
    "litre",
    "livestock unit",
    "megajoule",
    "meter",
    "meter-year",
    "person kilometer",
    "pig place",
    "square meter",
    "square meter-year",
    "ton kilometer",
    "unit",
    "vehicle kilometer",
}


def get_unusual_units_migration_data():
    """Only convert units that are not used in ecoinvent at all"""
    return {
        "fields": ["unit"],
        "data": [
            ((from_unit,), {"unit": to_unit, "multiplier": multiplier})
            for from_unit, to_unit, multiplier in DEFAULT_UNITS_CONVERSION
            if from_unit not in _USED_IN_ECOINVENT
        ],
    }
